File: Patterns/CreationMode/prototype_mode/src/com/fausto/poor/Sheep.java
package com.fausto.poor;

/**
 * @author: fausto
 * @date: 2021/10/29 15:20
 * @description:
 */
public class Sheep {
	String name;
	Integer age;
	String color;
	
	public Sheep(String name, Integer age, String color) {
		this.name = name;
		this.age = age;
		this.color = color;
	}
	
	@Override
	public String toString() {
		return "Sheep{" +
				"name='" + name + '\'' +
				", age=" + age +
				", color='" + color + '\'' +
				'}';
	}
}

File: Patterns/CreationMode/prototype_mode/src/com/fausto/shallowcopy/Sheep.kt
package com.fausto.shallowcopy

/**
 * @author: fausto
 * @date: 2021/10/29 16:16
 * @description:    浅拷贝: 只拷贝引用(指针,内存地址)而不拷贝内容(指针内容)
 */
data class Sheep(
    var name: String,
    var age: Int,
    var color: String,
    var friend: Sheep?
)

fun main() {
    val tom = Sheep("tom" , 2 , "white" , null)
    val jerry = Sheep("jerry" , 1 , "black" , tom)

    println(jerry.copy().apply { friend?.name = "mark" })
    println(jerry.copy())  //修改引用类型,会影响到所有指向该引用的对象

}